File: scripts/convert-folder.rb
require 'pathname'

# Open src folder and get list of music files
SRC_ROOT_DIR = File.expand_path(File.join(__dir__, '../music'))
DEST_ROOT_DIR = File.expand_path(File.join(__dir__, '../public_html/media/music'))

Dir.glob("#{SRC_ROOT_DIR}/**/*").each do |file|
    if File.directory?(file)
        next
    end

    extension = File.extname(file)
    
    unless ['.wav', '.mp3', '.m4a', '.aif'].include? extension
        next
    end

    filename = file.chomp(extension)
    # get directory relative to SRC_ROOT
    parent_dir = filename.reverse.chomp(SRC_ROOT_DIR.reverse).reverse

    filename_base_dest = File.expand_path(File.join(DEST_ROOT_DIR, parent_dir))
    parent_dir_dest = File.dirname(filename_base_dest)

    puts file

    `mkdir -p "#{parent_dir_dest}"`

    unless File.exist?("#{filename_base_dest}.opus")
        opus_bitrate = extension == '.wav' ? '128k' : '96k'
        `ffmpeg -i "#{file}" -c:a libopus -b:a #{opus_bitrate} -vbr:a on -strict -2 -y "#{filename_base_dest}.opus"`
    end

    if extension === '.wav' or extension === '.aif'
        # if .wav convert to aac as fallback
        unless File.exist?("#{filename_base_dest}.m4a")
            `ffmpeg -i "#{file}" -c:a aac -b:a 196k -y "#{filename_base_dest}.m4a"`
        end
    else
        # copy .mp3 and .m4a to destination as a fallback
        unless File.exist?("#{filename_base_dest}#{extension}")
            `cp "#{file}" "#{filename_base_dest}#{extension}"`
        end
    end
end

File: scripts/convert-folder.exs
defmodule SourceFile do
  defstruct [:path, targets: []]
end

defmodule ConvertFolder do
  @moduledoc """
  Converts wav or mp3 in source folder to opus or m4a in public html folder, if they don't already exist
  """
  defp source_dir do
    Path.join([__DIR__, "..", "music"]) |> Path.expand()
  end

  defp target_dir do
    Path.join([__DIR__, "..", "public_html", "media", "music"]) |> Path.expand()
  end

  defp get_files_recursive(source_path, %MapSet{} = extension_set) do
    case File.dir?(source_path) do
      true ->
        File.ls!(source_path)
        |> Enum.map(fn path ->
          Path.join(source_path, path) |> get_files_recursive(extension_set)
        end)
        |> Enum.flat_map(&Function.identity/1)

      false ->
        case MapSet.member?(extension_set, Path.extname(source_path)) do
          true -> [source_path]
          false -> []
        end
    end
  end

  defp to_source_file(path, %MapSet{} = target_files_set) do
    relative_path_without_extension = Path.relative_to(path, source_dir()) |> Path.rootname()

    targets =
      case Path.extname(path) == ".mp3" do
        false ->
          ["#{relative_path_without_extension}.m4a", "#{relative_path_without_extension}.opus"]

        true ->
          ["#{relative_path_without_extension}.mp3"]
      end
      |> Enum.filter(fn name -> !MapSet.member?(target_files_set, name) end)

    %SourceFile{path: path, targets: targets}
  end

  def main do
    # used to find out what target files we need to create
    target_files_set =
      get_files_recursive(target_dir(), MapSet.new([".m4a", ".opus", ".mp3"]))
      |> Enum.map(fn path -> Path.relative_to(path, target_dir()) end)
      |> MapSet.new()

    # get a list of source files, and the targets that need to be created from them
    source_files =
      get_files_recursive(source_dir(), MapSet.new([".wav", ".mp3", ".m4a", ".aif"]))
      |> Enum.map(fn path -> to_source_file(path, target_files_set) end)
      |> Enum.filter(fn source_file -> !Enum.empty?(source_file.targets) end)

    # get unique target folders to create, and create with mkdir_p
    target_directories_to_create =
      Enum.map(source_files, fn source_file -> source_file.targets end)
      |> Enum.flat_map(fn targets -> Enum.map(targets, &Path.dirname/1) end)
      |> MapSet.new()
      |> MapSet.to_list()
      |> Enum.filter(fn path -> !File.exists?(path) end)

    target_directories_to_create_count = Enum.count(target_directories_to_create)
    IO.puts("Creating #{target_directories_to_create_count} folders in #{target_dir()}\n")

    for target_directory <- target_directories_to_create do
      IO.puts("Creating #{target_directory}")
      File.mkdir_p!(target_directory)
    end

    source_files_count = Enum.count(source_files)
    # IO.inspect(source_files)
    IO.puts("\n#{source_files_count} files to convert")

    # https://hexdocs.pm/elixir/1.12/Task.html#async_stream/5
    # convert with Task.async_stream |> Enum.to_list()
  end
end

ConvertFolder.main()
